### PR TITLE
`map_single` is an object

### PR DESCRIPTION
**Pull request by eddie-c-davis**
## Purpose

This PR refactors the `map_single.compute` method into a class called `MapSingle` and `map_single.do_lagrangian_contributions` into the `LagrangianContributions` class.

## Code changes:

- Add `MapSingle` class
- Add `LagrangianContributions` class
- Remove `transliterated` (Python-only) implementation
- Update calling classes
- Add cached stencils in remaining procedural code
- Moves code in `mapn_tracer_2d` and `remap_profile` to init time.

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [x] Docstrings and type hints are added to new and updated routines, as appropriate

### CODE REVIEW

File: fv3core/stencils/map_single.py [hunk 1]
```diff
@@ -93,257 +90,175 @@ def lagrangian_contributions(
...
+
+        self.dp1 = utils.make_storage_from_shape(shape, origin=origin)
+        self.q4_1 = utils.make_storage_from_shape(shape, origin=origin)
+        self.q4_2 = utils.make_storage_from_shape(shape, origin=origin)
+        self.q4_3 = utils.make_storage_from_shape(shape, origin=origin)
+        self.q4_4 = utils.make_storage_from_shape(shape, origin=origin)
+
+        self.lagrangian_contributions = LagrangianContributions()
+
+        self._used_kords: Tuple[int] = (namelist.kord_tm, namelist.kord_tr)
```
Comment (eddie-c-davis) [inline fv3core/stencils/map_single.py:181]: Can these by not hardcoded?

File: fv3core/stencils/map_single.py [hunk 2]
```diff
@@ -93,257 +90,175 @@ def lagrangian_contributions(
...
+        qmin: float = 0.0,
+    ):
+        """
+        Compute x-flux using the PPM method.
+
+        Args:
+            q1 (in): Transported scalar
+            pe1 (in): ???
+            pe2 (out): ???
+            qs (out): ???
```
Comment (oelbert) [inline fv3core/stencils/map_single.py:213]: qs is the field to be remapped on the deformed grid, while q1 is the remapped field on the Eulerian grid.

It probably makes sense to rename these variables for clarity, actually

File: fv3core/stencils/map_single.py [hunk 3]
```diff
@@ -93,257 +90,175 @@ def lagrangian_contributions(
...
-        j2,
-        kord,
-        origin,
-        domain,
-        version,
-    )
-    return q1
+class LagrangianContributions:
+    """
+    Fortran name is lagrangian_contributions
```
Comment (oelbert) [inline fv3core/stencils/map_single.py:95]: There isn't really a Fortran equivalent to this class, it's just the mechanics of remap_z, map1_ppm, map_scalar, and mapn_tracer, reordered to fit in a gt4py stencil

File: fv3core/stencils/map_single.py [hunk 4]
```diff
@@ -93,257 +90,175 @@ def lagrangian_contributions(
...
-                                            - q4_2[ii, j, mm]
-                                            + q4_4[ii, j, mm] * (1.0 - r23 * esl)
-                                        )
-                                    )
-                                    k0 = mm
-                                    flag = 1
-                                    break
-                            # Add everything up and divide by the pressure difference
-                            q1[ii, j, k2] = qsum / (pe2[ii, j, k2 + 1] - pe2[ii, j, k2])
-                            break
```
Comment (oelbert) [inline fv3core/stencils/map_single.py:349]: really happy to be rid of this

File: tests/translate/translate_map_scalar_2d.py [hunk 5]
```diff
@@ -39,8 +39,7 @@ def compute(self, inputs):
         )
         inputs["i1"] = self.is_
         inputs["i2"] = self.ie
-        inputs["j1"] = inputs["j_2d"]
-        inputs["j2"] = inputs["j_2d"]
+        inputs["j1"] = inputs["j2"] = inputs["j_2d"]
```
Comment (oelbert) [inline tests/translate/translate_map_scalar_2d.py:42]: Good catch!

File: fv3core/stencils/mapn_tracer.py [hunk 6]
```diff
@@ -53,64 +52,55 @@ def compute(
...
             i2,
             j1,
             j2,
-            kord,
-            origin,
-            domain,
-            version,
+            map_single.origin,
+            map_single.domain,
         )
```
Comment (oelbert) [inline fv3core/stencils/mapn_tracer.py:101]: This is maybe more of a comment for the mapn_tracer oopification, but I think we could also get away with replacing the stuff above here in the for-loop with a call to MapSingle.

Again, completely fine for that to be a subsequent refactor!

**Review by oelbert — CHANGES_REQUESTED** [review]
Some minor points on the docstring and the kord/mode handling

File: fv3core/stencils/mapn_tracer.py (hunk 6 above)
Comment (eddie-c-davis) [reply fv3core/stencils/mapn_tracer.py:90]: Moved the origin and domain into the stencil.

File: fv3core/stencils/map_single.py (hunk 3 above)
Comment (eddie-c-davis) [reply fv3core/stencils/map_single.py:94]: Changed, thank you!

File: fv3core/stencils/map_single.py (hunk 2 above)
Comment (eddie-c-davis) [reply fv3core/stencils/map_single.py:205]: Awesome, thank you.

File: fv3core/stencils/map_single.py [hunk 7]
```diff
@@ -93,257 +90,175 @@ def lagrangian_contributions(
...
+        q4_2: FloatField,
+        q4_3: FloatField,
+        q4_4: FloatField,
+        dp1: FloatField,
+        i1: int,
+        i2: int,
+        j1: int,
+        j2: int,
+        origin: Tuple[int, int, int],
+        domain: Tuple[int, int, int],
```
[the remark this answers is not in the document]
Comment (eddie-c-davis) [reply fv3core/stencils/map_single.py]: Moved these into `__init__`.

File: fv3core/stencils/remap_profile.py [hunk 8]
```diff
@@ -509,55 +509,89 @@ class RemapProfile:
...
 
-    def __init__(self, kord, iv):
+    def __init__(
+        self,
+        kord: int,
+        iv: int,
+        i1: int,
+        i2: int,
+        j1: int,
+        j2: int,
```
Comment (oelbert) [inline fv3core/stencils/remap_profile.py:519]: Shoot, I completely forgot you'd have to propagate this to the RemapProfile class! Sorry to ask for so much extra work!

**Review by oelbert — APPROVED** [review]
This looks great! Thanks for the revisions!

**Pull request merged**